### PR TITLE
School Info Completeness - make SchoolInfo read-only

### PR DESCRIPTION
**Pull request by nkiruka**
The goal of over-riding the SchoolInfo update method is to prevent school info objects from being updated incorrectly.  This is accomplished by making SchoolInfo read-only.

### CODE REVIEW

**Comment by nkiruka** [comment]
Is it necessary to have the before_update hook?

Since read-only method is built into Rails, I am not sure if it is necessary to specify that the read-only method should be run only for updates.

Except from stack-overflow:
Looking at ActiveRecord::Persistence, everything ends up calling create_or_update behind the scenes.
```
def create_or_update
  raise ReadOnlyRecord if readonly?
  result = new_record? ? create : update
  result != false
end
So! Just:

def readonly?
  !new_record?
end
```

**Comment by nkiruka** [comment]
Another neat way to handle over-riding the SchoolInfo update method is using persist:
```
 def readonly?
    persisted?
  end
```
https://apidock.com/rails/v5.2.3/ActiveRecord/Persistence/new_record%3F

persisted? also checks if the record has not been destroyed.

**Review by bethanyaconnor — APPROVED** [review]

File: dashboard/app/models/school_info.rb [hunk 1]
```diff
@@ -82,6 +82,13 @@ def school_state=(input)
   # Only sync from school on create to avoid unintended updates to old data
   before_validation :sync_from_schools, on: :create
 
+  before_validation :readonly?, on: :update
```
Comment (joshlory) [inline dashboard/app/models/school_info.rb]: I'm not sure this is needed. The `readonly?` method below doesn't actually throw, it overrides the default [`ActiveRecord::Core#readonly?`](https://apidock.com/rails/ActiveRecord/Core/readonly%3F) method called in the code sample you linked.

The test passes for me without this line. Additionally, `readonly?` returns true in the update case, which would be the reverse of what we want. If the `before_validation` callback returned a false value it _used_ to halt the callback chain, but [that's no longer true in Rail 5](https://blog.bigbinary.com/2016/02/13/rails-5-does-not-halt-callback-chain-when-false-is-returned.html).

**Comment by joshlory** [comment]
> Is it necessary to have the before_update hook?

I don't think so, see my comment.

File: dashboard/app/models/school_info.rb (hunk 1 above)
Comment (nkiruka) [reply dashboard/app/models/school_info.rb]: @joshlory  I thought we want SchoolInfo to be readonly in update case so the data isn't altered.

**Review by bethanyaconnor — APPROVED** [review]

File: dashboard/app/models/school_info.rb (hunk 1 above)
Comment (joshlory) [reply dashboard/app/models/school_info.rb]: Correct! Looks like this is fixed with commit 828a16b9fe2da71ea8c038f5ad75b7725a9023fc.

**Pull request merged**